### ISSUE TITLE
No error reported for nonsensical assignment to a generic type's generic property type

### ISSUE THREAD

**Issue by aweebit** [microsoft/TypeScript#62665]
### 🔎 Search Terms

is not assignable to type error generic property infer

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries including "property" and "not assignable"

### ⏯ Playground Link

https://www.typescriptlang.org/play/?#code/C4TwDgpgBAygxgCwgWwIYB4AqA+KBeKAbyk3AgC4SoBfAbgCh7RIoBJAOwDMIAneJNKUjoYUCAA9gEdgBMAzrEQoMqdiGy4CoiVNkL+y9AEsuvErgD8VSuwgA3Xg3qcAruzjAjAe3ZQZKLxExSWl5RQEVNQ0ACgBKInooJKg4HzlgKFRKDm4+JUEyEU0oAEYAJgBmWigAehqoAFEeHi8eBVQFCUgPCBlE5NT2dKgAI0oYAG0AciEIKYBdfFLK6rrYBC8XABsZTK25LzFm1oAaUZcMmS8IOXYp4ABCemp6IA

### 💻 Code

```ts
type Schema<T> = { Type: T };

type InferSchemaType<S extends Schema<any>> = S extends Schema<infer T> ? T : never;

function demo<S extends Schema<any>>() {
    const a: InferSchemaType<S> = 123; // Errors as expected
    const b: S['Type'] = 123; // Should also error, but doesn't!
}
```

### 🙁 Actual behavior

No error is reported for the assignment to `b`.

### 🙂 Expected behavior

The following error is reported (just like for the assignment to `a`):

```
Type 'number' is not assignable to type 'S["Type"]'.
```

### Additional information about the issue

I discovered the bug while working on a library that is built on top of [Effect Schema](https://effect.website/docs/schema/introduction/). I now have to replace `S['Type']` with `Schema.Schema.Type<S>` everywhere, but I don't see any reason why that should be required and why TypeScript cannot report errors for `S['Type']` just as well.

**Comment by RyanCavanaugh** [comment]
🤖 Thank you for your issue! I've done some analysis to help get you started. This response is automatically generated; feel free to 👍 or 👎 this comment according to its usefulness.

## Similar Issues

Here are the most similar issues I found
- (68%) [microsoft/typescript#55383](https://redirect.github.com/microsoft/typescript/issues/55383): Generic indexed access on a substitution type with an array constraint resolves incorrectly
- (67%) [microsoft/typescript#26274](https://redirect.github.com/microsoft/typescript/issues/26274): Indexed Accesses Checking Allow Unsafe Assignments
- (65%) [microsoft/typescript#56388](https://redirect.github.com/microsoft/typescript/issues/56388): What is the type of T\['x'\] when T is type param of a generic function?

If your issue is a duplicate of one of these, feel free to close this issue. Otherwise, no action is needed.

**Comment by RyanCavanaugh** [comment]
Lookup types (`T['p']`) are eagerly resolved against the constraint of `T`. This is useful since there are two possible behaviors (eager or deferred) and two possible ways to get this info (lookup type or conditional type)

**Comment by aweebit** [comment]
@RyanCavanaugh this is quite confusing though. After all, when I write `S['Type']`, I am explicitly referring to the property of `S`, and not of its constraint, otherwise I would just write `Schema<any>['Type']`, and that's quite straightforward considering I am the one who defined that constraint.

By contrast, the property type is something I don't have any control over if I'm using a third-party library. Imagine it exposes something like this:

```ts
export type ThirdParty<T> = {
  p: SomeComplexTypeConversion<T>;
};
```

If `SomeComplexTypeConversion` is not exported along with `ThirdParty`, how do I access the actual type of `T['p']` where `T` is constrained by `ThirdParty<any>`? There currently doesn't seem to be a way to do it at all. Are library authors supposed to be aware of this and plan their exports accordingly so that for example `SomeComplexTypeConversion` is also exported? I don't see it happening – this is just too much to think of.

Also, after I replaced `S['Type']` with `Schema.Schema.Type<S>` everywhere in the code of my library built on top of [Effect Schema](https://effect.website/docs/schema/introduction/) today, I discovered quite a few mistakes in how my types were designed that had been hidden from me because of this eager resolution to `any`. Does this behavior have any advantages at all? Do you have an example where it could be useful? Because to me, it feels extremely confusing, not well-documented and hard to reason about.

**Comment by aweebit** [comment]
@RyanCavanaugh I don't think this should have been closed without a discussion.

**Comment by RyanCavanaugh** [comment]
@aweebit What do you want me to do about it?

**Comment by aweebit** [comment]
@RyanCavanaugh Is this behavior something that the TypeScript team is not willing to reconsider under any circumstances? Unless it is, it would be great if the issue remained open so that other developers could share their opinions, too.

Also, I would be grateful if you could answer my previous comment where I went into detail on what issues I see with this current behavior.

**Comment by RyanCavanaugh** [comment]
It's very unlikely we'd take an extremely large breaking change like that without a much clearer upside. There are reasons you might want the constrained-resolved type, and reasons you might want the deferred type, so making 2 syntaxes do 1 same thing implies the need for a third syntax, which then just opens the door to another "why doesn't this new one do the same thing as the other two". You're free to open a suggestion if you like be please also see https://github.com/Microsoft/TypeScript/wiki/FAQ#this-is-closed-but-should-be-open-or-vice-versa

**Comment by aweebit** [comment]
@RyanCavanaugh could you please tell me how to make this snippet type-safe?

```ts
type Schema<A, I> = {
    decode: (i: I) => A;
    encode: (a: A) => I;
};

function demo<Schemas extends Schema<any, any>[]>(schemas: Schemas) {
    return <S extends Schemas[number]>(schema: S) => {
        schema.decode(123); // should error!
        schema.encode(123); // should error!
    };
}
```

As far as I understand, the problem here is again that eager resolution of a lookup type. This time, it is the `Schemas[number]` type used as a constraint for another type parameter (`S`). The eager resolution results in `S` being basically equivalent to `Schema<any, any>`, so pretty much nothing we do with the `schema` parameter is type-safe, and that is of course not what we want.

So what would be the right inference trick here? Would it be this monster that ChatGPT suggest?

```ts
type TupleElements<T extends readonly unknown[]> =
  T extends readonly [infer Head, ...infer Tail]
    ? Head | TupleElements<Tail>
    : never;
```

Having to use something like this instead of a simple `Schemas[number]` is just terribly confusing, and probably also hurts performance.

And the worst part is, again, that developers might not realize at all that their code is unsafe and that those inference tricks have to be used to fix that because they just don't see any errors, and so they think there is no problem.

Aren't those reasons enough to at least consider changing the behavior? I really fail to see what upsides keeping it the way it is has. Even the fact that it would be a breaking change, and those should ideally be avoided, doesn't convince me. I still haven't seen an example where this eager resolution is exactly what the developer wants, and even if they exist, I don't see why the lookup cannot simply be done directly on the constraint type. I'm pretty sure 99% of the new errors this change would cause would actually be in places where there are currently hidden bugs that developers have missed because the errors were not there.

**Comment by RyanCavanaugh** [comment]
> Aren't those reasons enough to at least consider changing the behavior? 

I don't see any evidence that your proposal hasn't been considered. Again, please log a new suggestion if you want to advocate for changing it and want others to weigh in too. This is a bug report form, not a "convince me" form, and this (at least for now) is not a bug.

**Comment by aweebit** [comment]
The `TupleElements` trick from my previous comment doesn't actually work, although I first thought it did because errors were reported after I replaced `extends Schemas[number]` with `extends TupleElements<Schemas>` _([playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAygxgCwgWwIYB4CCAaKBJAPigF4oBvAKCmqgBMI4B7egLigAoBLNvAShKKYA3FRoQAdk1YdUbTP2JE8IgL4iKoSFAAqAVzAAbCAFEjyCcADO6bVAgAPYBNqWoAJwipajcQZBRdcQBrcUYAd3EAbQBdImJRWwcncRd3T29ff0jOcQAzCDcoAAl03AA6Cpz8wu1UTgNo0WoAfmL0qAAfHX0jUxQLa1r6giaoNnEIADcC9VzAuGBOHzoURnR4JDRXJOdXDZQMVHEQXCOQAhiCdktEA8s2fa3+Sho04F03cSh1u0dd7sMJjMA3Wty2BCuN02slgCiIL1eNChBzK9CkEHYAEYAEwAZl4QigAHoiVBLAhGLoDLQoAAjaBHOxuNyMQqoWmU4BQMCoNyocxOQqaaDITiWNDARBQHKWJxeACEo1eyLQZQk6KxeIJxNJ5Mp1LpDK+BRZbI5ui5PL5AoKUGFUFF4tQkoQ0vEsvSiteagoKgoQA))._

As it turns out, the errors were actually not about the type mismatch between the parameters and the arguments like they should be, but about the `decode` / `encode` properties not existing on `S`:

```
Property 'decode' does not exist on type 'S'. (2339)
Property 'encode' does not exist on type 'S'. (2339)
```

Apparently, TypeScript doesn't even realize that `TupleElements<Schemas>` is still a union of types constrained by `Schema<any, any>`, and that makes this `TupleElements` "solution" completely unusable.

Strangely enough, when I removed `readonly` from the definition of `TupleElements`, the errors were gone, but `S` was again reduced to just `Schema<any, any>` _([playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAygxgCwgWwIYB4CCAaKBJAPigF4oBvAKCmqgBMI4B7egLigAoBLNvAShKKYA3FRoQAdk1YdUbTP2JE8IgL4iKoSFAAqAVzAAbCAFEjyCcADO6bVAgAPYBNqWou8QGtxjAO7iA2gC6RMSitg5O4i5Q-pziAGYQAE5QABIQqLS4AHS5cYkp2qicBoGi1AD8aRm0UAA+OvpGpigW1kUlBOVQbOIQAG7J6vHucMCcjOJ0KIzo8EhorhHOrvMoGKjiILibIARBBOyWiOuWbGuL-JQ0UEkQwLpJU3N2jiuNhiZmbXMniwSHY4LWSwBREa43GhA9bZehSCDsACMACYAMy8IRQAD0WKglgQjF0BlqySSjCSAEJujdoWhshJ4Ui0Rjsbj8YTiXYkmTKd01BQVBQgA))._ So to summarize, it seems that eager type resolution sometimes happens even when inference tricks are used instead of lookup types, and in such cases, there doesn't seem to be a way to write type-safe code at all. I personally see a big problem with this.

@RyanCavanaugh I was hesitant about opening a new issue because usually, opening multiple issues about the same problem is frowned upon. Also, I don't have an alternative syntax proposal, but would like the semantics of lookup types to be changed in a way that can break things. But if you don't mind, I will create a new issue with this suggestion.
